src/charm.py

```diff
@@ -254,22 +259,29 @@ def _start(self, event: EventBase) -> None:
 
     def _restart(self, event: EventBase) -> None:
         """Handler for emitted restart events."""
-        # if not self.state.stable or not self.upgrade_events.idle:
-        #     event.defer()
-        #     return
         if not self.state.unit_server.started:
             self.reconcile(event)
             return
 
         logger.info(f"{self.unit.name} restarting...")
         self.workload.restart()
 
+        # Allow the service to start up safely on the snap level
         start_time = time.time()
         while not self.workload.alive() and time.time() - start_time < RESTART_TIMEOUT:
             time.sleep(5)
 
+        # Allow the service to establish
+        # Reason: we are emitting an 'update-status' right after
+        # If the service is not yet functional, the status is set as
+        # 'Service unavailable' until the next 'update-status' hook execution
+        start_time = time.time()
+        unit_healthy, _ = self.health_manager.unit_healthy()
+        while not unit_healthy and time.time() - start_time < SERVICE_AVAILABLE_TIMEOUT:
+            time.sleep(5)
+            unit_healthy, _ = self.health_manager.unit_healthy()
+
         clear_status(self.unit, [MSG_STARTING, MSG_STARTING_SERVER])
-        self.on.update_status.emit()
```

Comment:
Avoid a loop here, as described on: https://github.com/canonical/opensearch-dashboards-operator/pull/114#pullrequestreview-2369694313

There is no point in speeding up an `update-status`, it will happen anyways. In this case, it is better to leave some time between this `_restart` call and the next `update-status` for the system to settle.

Reply:
FYI, the proposal above reintroduces the risk of continuous restart loop if the unit is failing because of some other problem...